server/src/internalClusterTest/java/org/opensearch/search/backpressure/SearchBackpressureIT.java

```diff
@@ -0,0 +1,284 @@
+                    // sleeping for the percentage of unladen time every 100ms
+                    if (System.currentTimeMillis() % 100 == 0) {
+                        Thread.sleep((long) Math.floor((1 - load) * 100));
+                    }
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        });
+    }
+
+    private void resetSettingValue(String key) {
+        client().admin().cluster().prepareUpdateSettings().setPersistentSettings(Settings.builder().putNull(key)).get();
+    }
+
+    public static class TestRequest extends ActionRequest {
+        Runnable runnable;
+
+        public TestRequest(StreamInput in) {}
+
+        public TestRequest(Runnable runnable) {
+            this.runnable = runnable;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+
+        @Override
+        public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
+            return new SearchShardTask(id, type, action, "", parentTaskId, headers);
+        }
+
+        public void execute() {
+            runnable.run();
+        }
+    }
+
+    public static class TestResponse extends ActionResponse {
+        public TestResponse() {}
+
+        public TestResponse(StreamInput in) {}
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {}
+    }
+
+    public static class TestTransportAction extends HandledTransportAction<TestRequest, TestResponse> {
+        public static final ActionType<TestResponse> ACTION = new ActionType<>("internal::test_action", TestResponse::new);
+
+        @Inject
+        public TestTransportAction(TransportService transportService, NodeClient client, ActionFilters actionFilters) {
+            super(ACTION.name(), transportService, actionFilters, TestRequest::new, ThreadPool.Names.GENERIC);
+        }
+
+        @Override
+        protected void doExecute(Task task, TestRequest request, ActionListener<TestResponse> listener) {
+            try {
+                SearchShardTask searchShardTask = (SearchShardTask) task;
+                // starting resource tracking for the current thread
+                ResourceUsageMetric[] initialTaskResourceMetrics = new ResourceUsageMetric[] {
+                    new ResourceUsageMetric(ResourceStats.MEMORY, 0),
+                    new ResourceUsageMetric(ResourceStats.CPU, 0) };
+                task.startThreadResourceTracking(
+                    Thread.currentThread().getId(),
+                    ResourceStatsType.WORKER_STATS,
+                    initialTaskResourceMetrics
+                );
+                long startTime = System.nanoTime();
+
+                // Doing a busy-wait until task cancellation or timeout.
+                do {
+                    request.execute();
+                    Thread.sleep(500);
```

Comment:
nit: Probably remove this sleep and let the `request.execute()` decide if it wants to sleep and for how long.

Reply:
Noted.